Guides 2.0 - Actions # Exception Handling (#118)
hanami/guides: content/v2.0/actions/exception-handling.md

---
title: Exception handling
order: 100
---

> When a request crashes with an exception, you want to handle it in a graceful manner. In order to do this, you can use the `handle_exception` action API.

Reviewer: ```suggestion
When a request crashes with an exception, you may want to handle it in a graceful manner. To do this, you can use `handle_exception` in your actions.
```

Actions do not handle exceptions raised by your application by default. To handle exceptions, you need to define how a specific exception type should be translated into an HTTP response.

An exception handler can be a HTTP status code (eg. `500`, `401`), or a symbol that represents a method on the action.

## Status codes

```ruby
# app/actions/books/index.rb
module Bookshelf
  module Actions
    module Books
      class Index < Bookshelf::Action
        handle_exception StandardError => 500

        def handle(request, response)
          raise "error"
        end
      end
    end
  end
end
```

> In the above action, when `StandardError` is raised in the `#handle` method, a `500 Internal Server Error` will be returned.

Reviewer: ```suggestion
In the above action, when `StandardError` is raised in the `#handle` method, a basic `500 Internal Server Error` will be returned.
```

![giphy](https://user-images.githubusercontent.com/3134/203051483-cf4de175-1166-493d-b095-47435502ca0c.gif)

Reviewer 2: lol @timriley

<p><img src="/v2.0/actions/default-error-response.png" alt="Default error response"></p>

## Custom method handlers

To do more with an exception than simply rendering a particular status code, call a method by providing a symbol with the method's name:

```ruby
module Bookshelf
  module Actions
    module Books
      class Index < Bookshelf::Action
        handle_exception RecordNotFound => 404
        handle_exception StandardError => :handle_standard_error

        def handle(*, response)
          raise "error"
        end

        private

        def handle_standard_error(request, response, exception)
          response.status = 500
          response.format = :json
          response.body = {error: "Sorry, something went wrong handling your request"}.to_json
        end
      end
    end
  end
end
```

> This time, when `StandardError` is raised, the `#handle_standard_error` will render a json response.

Reviewer: ```suggestion
Here, when `StandardError` is raised, `#handle_standard_error` will prepare a JSON response.
```

---
title: Exception Handling
order: 100
---

> Actions have an elegant API for exception handling. You can define the exception handling per action, using app/slice configuration.

Reviewer 3: ```suggestion
When a request crashes with an exception, you want to handle it in a graceful manner. In order to do this, you can use the `handle_exception` action API.
```

## Default behavior

> By default, actions do not handle errors risen by the app, and to handle them, you need to define the error mapping to the handler.

Reviewer 3: ```suggestion
By default, actions do not handle exceptions raised by the app. To handle them, you need to define how a specific exception type should be translated into an HTTP response.
```

An exception handler can be a valid HTTP status code (eg. `500`, `401`), or a `Symbol` that represents an action method.

```ruby
# app/actions/books/index.rb
module Bookshelf
  module Actions
    module Books
      class Index < Bookshelf::Action
        handle_exception StandardError => 500

        def handle(*, response)
          raise
        end
      end
    end
  end
end
```

`handle_exception` accepts a Hash where the key is the exception to handle, and the value is the corresponding HTTP status code.

> In our example, when `StandardError` is raised, it will be handled as a `500` (Server Error).

Reviewer 3: ```suggestion
In our example, when `StandardError` is raised, it will be handled as a `500` (Server Error) response.
```

![[default-error-response.png]]

## Custom error handlers

> If the mapping with a custom HTTP status doesn't fit our needs, we can specify a custom handler and manage the exception by ourselves.

Reviewer 3: ```suggestion
If the mapping with a custom HTTP status doesn't fit your needs, you can specify a custom handler and manage the exception by yourself.
```

If you'll pass a symbol instead of the HTTP status code, the action will look for the method with the same name.

```ruby
module Bookshelf
  module Actions
    module Books
      class Index < Bookshelf::Action
        handle_exception StandardError => :my_error_handler

        def handle(*, response)
          raise
        end

        private

>         def my_error_handler(req, res, exception)

Reviewer 3: ```suggestion
        def my_error_handler(request, response, exception)
```

          res.status = 500
          res.body = "You've found a Unicorn!"
        end
      end
    end
  end
end
```

![[customized-error-response.png]]

In the example above we changed the default message delivered to the client.

> This time, when `StandardError` will be raised, it'll be handled by our method.

Reviewer 3: ```suggestion
This time, when `StandardError` is raised, it'll be handled by our method.
```
